// @solana/actions/lib/cjs/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ACTIONS_CORS_HEADERS_MIDDLEWARE = exports.ACTIONS_CORS_HEADERS = exports.BLOCKCHAIN_IDS = exports.BLINKS_QUERY_PARAM = exports.MEMO_PROGRAM_ID = exports.HTTPS_PROTOCOL = void 0;
/** @internal */
exports.HTTPS_PROTOCOL = "https:";
/** Program Id for the SPL Memo program */
exports.MEMO_PROGRAM_ID = "MemoSq4gqABAXKb96qnH8TysNcWxMyWCqXgDLGmfcHr";
/**
 * Specification declared URL query parameter to detect blinks via interstitial website URLs.
 *
 * Note: The `action` query parameter should begin with the `solana-action:` protocol identifier
 *
 * Example: `https://dial.to/?action=solana-action:https://jupiter.dial.to/swap/SOL-Bonk`
 */
exports.BLINKS_QUERY_PARAM = "action";
/**
 * Blockchain IDs for Solana from CAIP
 *
 * @see https://namespaces.chainagnostic.org/solana/caip10
 */
exports.BLOCKCHAIN_IDS = {
    mainnet: "solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp",
    devnet: "solana:EtWTRABZaYq6iMfeYKouRu166VU2xqa1",
    testnet: "solana:4uhcVJyU9pJkvQyS88uRDiswHXSCkY3z",
};
/**
 * Standard headers for use within frameworks that use the native `HeadersInit` (like NextJS)
 *
 * Note: `Access-Control-Allow-Origin=*` should ONLY be set on your Actions API routes and `actions.json`.
 * Setting "allow origin to any" on other routes on your server is bad practice and should be avoided.
 */
exports.ACTIONS_CORS_HEADERS = {
    "Access-Control-Allow-Origin": "*",
    "Access-Control-Allow-Methods": "GET,POST,PUT,OPTIONS",
    "Access-Control-Allow-Headers": "Content-Type, Authorization, Content-Encoding, Accept-Encoding, X-Accept-Action-Version, X-Accept-Blockchain-Ids",
    "Access-Control-Expose-Headers": "X-Action-Version, X-Blockchain-Ids",
    "Content-Type": "application/json",
};
/**
 * Standard headers for use within frameworks that use middleware to handle CORS headers
 * (like Hono, Express, and Fastify)
 *
 * Note: `origin=*` should ONLY be set on your Actions API routes and `actions.json`.
 * Setting "allow origin to any" on other routes on your server is bad practice and should be avoided.
 */
exports.ACTIONS_CORS_HEADERS_MIDDLEWARE = {
    origin: "*",
    methods: ["GET", "POST", "PUT", "OPTIONS"],
    allowedHeaders: [
        "Content-Type",
        "Authorization",
        "Content-Encoding",
        "Accept-Encoding",
        "X-Accept-Action-Version",
        "X-Accept-Blockchain-Ids",
    ],
    exposedHeaders: [
        "X-Action-Version",
        "X-Blockchain-Ids",
    ],
};
